networkgen/mapgen.py

```diff
@@ -0,0 +1,447 @@
+        """
+
+        lignames = [self.lig_dict[k]['Name'] for k in self.lig_dict.keys()]
+        lignames = [lig for charge_group in lignames for lig in charge_group] #flatten ligand names
+
+        mols = [self.lig_dict[k]['Mol'] for k in self.lig_dict.keys()]
+        mols = [m for charge_group in mols for m in charge_group] #flatten mols
+
+        if len(lignames) < 2:
+            raise Exception("Number of ligands ({}) should be > 1".format(len(lignames)))
+
+        ################## RDKIT IMAGE GENERATION ################
+        # flatten all ligands into 2D space. Note that most rdkit mol operations are in-place.
+        for m in mols:
+            rdDepictor.Compute2DCoords(m)
+            m.UpdatePropertyCache()
+
+
+        # find the core using standard MCS algorithm. Flatten the core again just to be sure.
+        mcs = rdFMCS.FindMCS(mols, matchValences=False,
+                                        ringMatchesRingOnly=True,
+                                        completeRingsOnly=True,
+                                        matchChiralTag=False)
+
+        core = Chem.MolFromSmarts(mcs.smartsString)
+        rdDepictor.Compute2DCoords(core)
+
+        # find subtructure matches with MCS per ligand, then tag matching atom indices in each ligand.
+        ps = Chem.AdjustQueryParameters.NoAdjustments()
+        ps.makeDummiesQueries=True
+        qcore = Chem.AdjustQueryProperties(core,ps)
+        #mhs = [Chem.AddHs(x,addCoords=True) for x in ms]
+        mms = [x for x in mols if x.HasSubstructMatch(qcore)]
+        for m in mms:
+            for atom in m.GetAtoms():
+                atom.SetIntProp("SourceAtomIdx",atom.GetIdx())
+
+        # do an RDKit R-group decomposition.	
+        groups,_ = rdRGroupDecomposition.RGroupDecompose([qcore],mms,asSmiles=False,asRows=True)
+
+        # call the writer function with each molecule.
+        for i, m in enumerate(mols):
+            png = self.generateImages(m,groups[i],qcore,lbls=('R1','R2','R3','R4', 'R5', 'R6', 'R7'),
+                                    legend=lignames[i],
+                                    width=400,height=400)
+
+    def set_ligdict(self):
+        for mol in self.suppl:
+            charge = Chem.rdmolops.GetFormalCharge(mol)
+            ligand = self.get_ligand(mol)
+            ligand.save()
+            v = self.lig_dict.setdefault(charge, {'Name': [], 'Mol': [], 'FP': []})
+            v['Name'].append(mol.GetProp('_Name'))
+            v['Mol'].append(mol)
+            if self.metric != g.MCS:
+                v['FP'].append(self.make_fp(mol))
+        self.writeLigandImages()
+
+    def sim_mx(self):
+        if self.metric in [g.Tanimoto, g.MFP]:
+            from rdkit import DataStructs
+        elif self.metric == g.MCS:
+            from rdkit.Chem import rdFMCS
+        elif self.metric == g.SMILES:
+            from Bio import pairwise2
```

Comment:
Biopython is not yet in requirements.txt

Reply:
Also, why not just make the imports at the top? and get rid of this "seemingly" unnecessary elif statements?

Reply:
I have the impression it's after some kind of optimization (not importing if not needed). Not sure if it's effective, never measured it.